src/MiddlewareFactory.php

```diff
@@ -0,0 +1,140 @@
+ * - pipeline() will create a MiddlewarePipe instance from the array of
+ *   middleware passed to it, after passing each first to prepare().
+ */
+class MiddlewareFactory
+{
+    /**
+     * @var MiddlewareContainer
+     */
+    private $container;
+
+    public function __construct(MiddlewareContainer $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * @param string|array|callable|MiddlewareInterface $middleware
+     * @throws Exception\InvalidMiddlewareException if argument is not one of
+     *    the specified types.
+     */
+    public function prepare($middleware) : MiddlewareInterface
+    {
+        if ($middleware instanceof MiddlewareInterface) {
+            return $middleware;
+        }
+
+        if (is_callable($middleware)) {
+            return $this->callable($middleware);
+        }
+
+        if (is_array($middleware)) {
+            return $this->pipeline(...$middleware);
+        }
+
+        if (! is_string($middleware) || ! $middleware) {
+            throw Exception\InvalidMiddlewareException::forMiddleware($middleware);
+        }
+
+        return $this->lazy($middleware);
+    }
+
+    /**
+     * Decorate callable standards-signature middleware via a CallableMiddlewareDecorator.
+     */
+    public function callable(callable $middleware) : CallableMiddlewareDecorator
+    {
+        return new CallableMiddlewareDecorator($middleware);
+    }
+
+    /**
+     * Create lazy loading middleware based on a service name.
+     */
+    public function lazy(string $middleware) : Middleware\LazyLoadingMiddleware
+    {
+        return new Middleware\LazyLoadingMiddleware($this->container, $middleware);
+    }
+
+    /**
+     * Create a middleware pipeline from an array of middleware.
+     *
+     * This method allows passing an array of middleware as either:
+     *
+     * - discrete arguments
+     * - an array of middleware, using the splat operator: pipeline(...$array)
+     * - an array of middleware as the sole argument: pipeline($array)
+     *
+     * Each item is passed to prepare() before being passed to the
+     * MiddlewarePipe instance the method returns.
+     *
+     * @param string|array|MiddlewarePipe $middleware
+     */
+    public function pipeline(...$middleware) : MiddlewarePipe
+    {
+        // Allow passing arrays of middleware or individual lists of middleware
+        if (is_array($middleware[0])
+            && count($middleware) === 1
+        ) {
+            $middleware = array_shift($middleware);
+        }
+
+        $pipeline = new MiddlewarePipe();
+        foreach ($middleware as $m) {
+            $pipeline->pipe($this->prepare($m));
+        }
+        return $pipeline;
+    }
+
+    /**
+     * Segregate one or more middleware by path.
+     *
+     * Creates and returns a PathMiddlewareDecorator instance after first
+     * passing the $middleware argument to prepare().
+     *
+     * @param string|array|MiddlewareInterface $middleware
```

Comment:
missing @param for first parameter

Reply:
This is standard across major IDEs, at this point, and phpdocumentor infers types from typehints as well. We've already decided for zend-coding-standard to require annotations only if they're adding information, or the type is unspecified.